### DISCUSSION TITLE
Features and Suggestions

### DISCUSSION THREAD

**Discussion by MrKuenning** [advplyr/audiobookshelf#195] [Ideas]
First of all, thank you to those working on this project.  I have been looking for a self-hosted audiobook library for a while now.  After having mixed results using PLEX as an audiobook server and trying others it's nice to see a new one with promise.

I am going to mention my observations on it but I am in no way trying to be negative or critical.  Merely sharing what I think may improve the overall project.

In no particular order:

** - Root Only Admin**
The first thing I did after installing was to create a new account for myself.  I made that account admin, but after logging in I noticed I did not have any admin rights.  I could not change settings, scan for books, or even see statistics.   

I am not sure if only the Root user can do these things, but I would like to be able to disable the root account and grant that level of rights to other accounts.  I also think all users should be able to see their statistics.

** - More Themes**
The wooden bookshelf has a neat rustic aesthetic.  But it is a very niche style.  Most services that I have used that have a natural-looking UI I have tended to turn off after a while because it made the UI busy or cluttered looking.  

I would suggest a few options like a dark wood and light wood option as well as some solid colors like a more modern grey theme.

** - Don't rely on Folder structure**
There seems to be an interesting mix of folder structure and metadata going on here. I don't mind some preset folder layouts.  I am used to Plex or Kodi using a single folder per movie or show.  But typically they only focus on the folder the file is in and not so much on the parent folder. Also for audio, the information that matters is in the file. This is particularly important for me with the author folder.  I don't sort my audiobooks by the author any more than I sort movies by the director.  Many of my series have mixed authors. So it would make for some awkward folder structure to have to split up a series in different directory trees.

I totally get why some people may want to sort books by the author.  But I think a good strategy would be to put less importance on the parent folders and more on the final folder and the metadata.  So if I sorted all my books folders as a child of a **series folder** or **topic folder** or **author folder** it doesn't matter.

Here is an Example:

Starwars -
	Legacy Era - 
		Series Name - 
			Book One Name -
				Series Name - 1 - Book Name.m4b			
			Book Two Name - 
				Series Name - 1 - Book Name.m4b	

Those two books may be written by different people and my understanding is that I would need to place them under subfolders belonging to those authors.

 **- Support M4b chapters in the UI**
Many of my books were MP3 files and I have converted them to an audiobook file by combining them into one M4b file and used the old MP3s as chapters.  This is a nice clean way to have a single file per book and still have chapters with names. I noticed that in the UI on the browser when I edit the book details it has a Chapters tab that lists out all of the chapter's names and time stamps.  However when I select a file to play there is no list of chapters to start from.  Most audiobook players will show a list of chapters and also show the current time status of the current chapter as well as the book as a whole.   

I would suggest you display a M4b file with chapters the same way you would display a folder full of MP3s.  When you tap on the book just list out all the chapter names and times and let users select which they want to play.  When playing a file I would suggest showing the chapter or file that is currently playing with the time through that file as well as time remaining on the book as a whole. Aditianly it would be nice if the currently listened to chapters or files are marked as listened to on the book info screen.

 **- Inconsistent Resume**
I played a book in the app and then paused it in the middle of the book and then hit the X to end the stream.
I then when back to the book and hit play and it started at the beginning again.  I tried pausing in the middle of the book again. and went back to the library. In the list view, it showed that I was 45% through the book.  So I went back into the book and hit play and again it started at the beginning.  I went back to the list view and this time I hit play on the play button in the list view.  This time is resumed at 45%.   

I am not sure if that is a feature or a bug.  But I found it confusing that playing from list view resumed it but playing from the book details screen started the book over again.

 **- Windows Support**
This is a personal preference.  But It would be nice to have windows install support.  I use many of the ***arr tools  Radarr, Sonarr, Reader, Prowlarr Etc.  They are all running on one windows server and access files off an SMB share.  Installing on Docker and mounting that share via CIFS is was a break in my process and may be a hurdle for non-Linux people.

 **- Focus on one feature set**
I noticed a lot of mention about including eBook or CBZ support.  And while I think that would be awesome,  I feel like the audio support has a ways to go and should really be nailed down first before bringing on completely new functions of the app.

Again I thank you all for working and assisting with this project.  I look forward to seeing it evolve.
Feel free to hit me up for any further explanation on any item listed above.

**Comment by advplyr** [comment]
Thank you for this in-depth analysis. This project has been evolving from feedback and suggestions since the beginning. The initial release was one page with a basic scanner and streaming, no accounts, no series, no settings, even the logo was a user contribution ( #13 ). I say that just to say that feedback and suggestions are greatly appreciated, to the smallest detail.


> ** - Root Only Admin** The first thing I did after installing was to create a new account for myself. I made that account admin, but after logging in I noticed I did not have any admin rights. I could not change settings, scan for books, or even see statistics.
> 
> I am not sure if only the Root user can do these things, but I would like to be able to disable the root account and grant that level of rights to other accounts. I also think all users should be able to see their statistics.


This is in the plans, an open issue exists for it already #137. I was going back and forth on whether root and admin should have all the same access, but it seems the consensus is that they should. The account types "admin", "user" and "guest" don't serve much purpose yet except for setting default permissions.
The stats page is still very young. The listening data is still not being recorded in android auto and offline on the app. It also needs a refactor to work better with libraries, but eventually it will be for all users. First mention and design #167 

> ** - More Themes** The wooden bookshelf has a neat rustic aesthetic. But it is a very niche style. Most services that I have used that have a natural-looking UI I have tended to turn off after a while because it made the UI busy or cluttered looking.
> 
> I would suggest a few options like a dark wood and light wood option as well as some solid colors like a more modern grey theme.

Definitely agree with this, it has been brought up a few times. In #186 I asked if some patterns could be suggested, I'll pass along that request to you if you have some options.

> ** - Don't rely on Folder structure** There seems to be an interesting mix of folder structure and metadata going on here. I don't mind some preset folder layouts. I am used to Plex or Kodi using a single folder per movie or show. But typically they only focus on the folder the file is in and not so much on the parent folder. Also for audio, the information that matters is in the file. This is particularly important for me with the author folder. I don't sort my audiobooks by the author any more than I sort movies by the director. Many of my series have mixed authors. So it would make for some awkward folder structure to have to split up a series in different directory trees.

The main reason I did this in the beginning was because it was the easiest way to pull in a lot of details without parsing ID3 tags or using external sources. I'm working on a new scanner now with the idea being that the order of precedence on how the book details are filled out should be configurable. The issue with series having different authors was mentioned the other day #191 and led to me starting the new scanner.

> **- Support M4b chapters in the UI** Many of my books were MP3 files and I have converted them to an audiobook file by combining them into one M4b file and used the old MP3s as chapters. This is a nice clean way to have a single file per book and still have chapters with names. I noticed that in the UI on the browser when I edit the book details it has a Chapters tab that lists out all of the chapter's names and time stamps. However when I select a file to play there is no list of chapters to start from. Most audiobook players will show a list of chapters and also show the current time status of the current chapter as well as the book as a whole.
> 
> I would suggest you display a M4b file with chapters the same way you would display a folder full of MP3s. When you tap on the book just list out all the chapter names and times and let users select which they want to play. When playing a file I would suggest showing the chapter or file that is currently playing with the time through that file as well as time remaining on the book as a whole. Aditianly it would be nice if the currently listened to chapters or files are marked as listened to on the book info screen.

I'm not sure what you mean about showing a list of chapters when playing the audiobook. There is a chapter icon on the audio player. @p1xelll designed the mobile app player which is closer to what you are describing https://github.com/advplyr/audiobookshelf-app/discussions/26. I think the web app player will be built to match that mobile design.

> **- Inconsistent Resume** I played a book in the app and then paused it in the middle of the book and then hit the X to end the stream. I then when back to the book and hit play and it started at the beginning again. I tried pausing in the middle of the book again. and went back to the library. In the list view, it showed that I was 45% through the book. So I went back into the book and hit play and again it started at the beginning. I went back to the list view and this time I hit play on the play button in the list view. This time is resumed at 45%.
> 
> I am not sure if that is a feature or a bug. But I found it confusing that playing from list view resumed it but playing from the book details screen started the book over again.

This definitely sounds like a bug, and a strange one because the list and grid view call the same play function. I will look into this.

> **- Windows Support** This is a personal preference. But It would be nice to have windows install support. I use many of the ***arr tools Radarr, Sonarr, Reader, Prowlarr Etc. They are all running on one windows server and access files off an SMB share. Installing on Docker and mounting that share via CIFS is was a break in my process and may be a hurdle for non-Linux people.

It just needs an installer to be built, I can build the executable but there is more work involved for the installer. I think this will be available in the not-so-far future.

> **- Focus on one feature set** I noticed a lot of mention about including eBook or CBZ support. And while I think that would be awesome, I feel like the audio support has a ways to go and should really be nailed down first before bringing on completely new functions of the app.

Agreed!
